ServerCore/Pages/Submissions/Index.cshtml.cs

```diff
@@ -31,26 +32,56 @@ public IndexModel(ServerCore.DataModel.PuzzleServerContext context)
 
         public async Task<IActionResult> OnPostAsync(int puzzleId, int teamId)
         {
+
+            // Don't allow submissions after the answer has been found.
+            if (AnswerToken != null)
+            {
+                return Page();
+            }
+
+            // TODO: Once auth exists, we need to check if the team has access
+            // to this puzzle.
+
             if (!ModelState.IsValid)
             {
                 return Page();
             }
 
+            await SetupContext(puzzleId, teamId);
+
             // Create submission and add it to list
-            Submission.TimeSubmitted = DateTime.Now;
-            Submission.Puzzle = await _context.Puzzles.SingleOrDefaultAsync(p => p.ID == puzzleId);
-            Submission.Team = await _context.Teams.Where((t) => t.ID == teamId).FirstOrDefaultAsync();
+            Submission.TimeSubmitted = DateTime.UtcNow;
+            Submission.Puzzle = PuzzleState.Puzzle;
+            Submission.Team = PuzzleState.Team;
+            Submission.Response = await _context.Responses.Where(
+                r => r.Puzzle.ID == puzzleId &&
+                     Submission.SubmissionText == r.SubmittedText)
+                .FirstOrDefaultAsync();
 
-            List<Response> responses = await _context.Responses.Where(r => r.Puzzle.ID == puzzleId && Submission.SubmissionText == r.SubmittedText).ToListAsync();
-            Submission.Response = responses.FirstOrDefault();
+            Submissions.Add(Submission);
 
             // Update puzzle state if submission was correct
             if (Submission.Response != null && Submission.Response.IsSolution)
             {
-                var statesQ = await PuzzleStateHelper.GetFullReadWriteQueryAsync(_context, this.Event, Submission.Puzzle, Submission.Team);
-                PuzzleStatePerTeam puzzleState = await statesQ.FirstOrDefaultAsync();
-                puzzleState.IsSolved = true;
-                Submission.TimeSubmitted = (DateTime)puzzleState.SolvedTime;
+                PuzzleState.IsSolved = true;
+                PuzzleState.SolvedTime = Submission.TimeSubmitted;
+            }
+            else if (Submission.Response == null)
+            {
+                // If the submission was incorrect and not a partial solution,
+                // we will do the lockout computations now.
+                double? lockoutTime = ComputeLockoutTime(Event,
+                                                         Submissions,
+                                                         PuzzleState);
+                if (lockoutTime != null)
+                {
+                    PuzzleState.SetPuzzleLockout((double)lockoutTime);
+                }
+                // We also determine if the puzzle should be set to email-only mode.
+                if (IsPuzzleSubmissionLimitReached(Event, Submissions))
```

Comment:
Compute this first - possible bad vibes if a single submission creates a lockout and also puts you in email mode? No need for a double penalty in this case.